accepted/2023/net8.0-browser-tfm.md

```diff
@@ -0,0 +1,187 @@
+
+We reasoned that we did not need a TFM as we could always provide a PAL over any functionality we needed and we would annotate the APIs with the `[SupportedOSPlatform]` attribute and an analyzer that would warn against incorrect usage when targeting the browser.
+
+Over time, we have learned that the lack of a TFM is limiting for us and our customers. The lack of a TFM introduces friction in two ways:
+* **Exposing additional customer-facing APIs when running in browser (and not on desktop)**: For example, Blazor includes many platform agnostic abstractions like IJSRuntime, IJSInProcessRuntime and different implementations to be able to reflect the different capabilities of the different platforms. This makes taking advantage of some of the webassembly unique capabilities more challenging as this kind of functionality needs to be carefully designed, and creates other problems like lack of discoverability (you need to know the existence of IJSInProcessRuntime and downcast to it).
+  ```mermaid
+  classDiagram
+    IJSRuntime <|-- IJSInProcessRuntime
+    class IJSRuntime {
+      <<interface>>
+      InvokeAsync<&zwj;TValue>(string identifier, object[] args) ValueTask~TValue~
+      InvokeAsync<&zwj;TValue>(string identifier, CancellationToken cancellationToken, object?[]? args) ValueTask~TValue~
+    }
+    class IJSInProcessRuntime {
+      <<interface>>
+      Invoke<&zwj;TResult>(string identifier, params object?[]? args) TResult
+    }
+  ```
+* **Varying framework/library code contents/behavior based on whether it's running in the browser or not:** For example, having DI dependencies or other library code references varying by platform. Doing it based on TFM (not on RuntimeInformation) helps compilation and trimming.
+  * A library that targets Blazor Server and Webassembly must provide at a minimum an additional package for the server to avoid bringing the server dependencies to webassembly.
+  * A webassembly app with a companion server backend must also have, a project to host the webassembly app, a project to share common abstractions and a project to host the server backend. This exact scheme is reflected in the default Blazor Webassembly Hosted template.
+
+One of the reasons we were hesitant to ask for a TFM in the initial release of Blazor was that we were concerned about people writing .NET webassembly specific code that did not work with Blazor Server. Over time, we have determined that it is not a real risk for two reasons:
+* Blazor Server has a healthy user base, which makes it desirable for library authors to target net8.0.
+* A core piece of Blazor functionality (prerendering) depends on being able to run the application on net8.0.
+
+Given this, we consider unlikely that developers will author wasm specific code (net8.0-wasm) without also including a net8.0 version, as it will make it incompatible with any other Blazor flavor.
```

Comment:
I think that people will create non-blazor-rendering client side C# components which have strong JavaScript dependency, which would not be able to run on server side. Should they use `net8.0-wasm` ?

Example would be C# wrapper for WebGL.

Reply:
If it can only run in the browser that would be fine. For running on regular .NET they could provide an alternative version that targets .NET 8.0 in that case.

Reply:
They could or they have to, to keep it Blazor compatible ?

Reply:
Only if they wanted to run on Blazor Server or do Server Side Rendering, but it would be reasonable for a standalone Blazor app

Reply:
> Right now, we don't have that limitation. We could have the same assembly to run on both client and server side and just have conditional code.

You can still have that decision made at runtime if you use net8.0, won't you? The only requirement is that the APIs you need are available as a net8.0 target.

The difference is that the decision gets made at runtime, not compile time, and you do not get to differentiate the set of dependencies.

Reply:
If you use `net8.0` you would not be able to use the APIs of other assemblies which target `net8.0-browser`, right ?

Reply:
> If you use net8.0 you would not be able to use the APIs of other assemblies which target net8.0-browser, right ?

I think "use" needs clarification here. Strictly speaking you won't be able to reference a package/project that targets `net8.0-wasm` from a `net8.0` project. Similarly speaking, if you author a library for `net8.0-wasm` only, it can be consumed by anyone targeting `net8.0`.

Now, if your app targets net8.0-wasm, it can consume libraries that target either net8.0 or net8.0-wasm, and if your app targets net8.0, then it won't be able to consume anything that exclusively targets net8.0-wasm.

From the application developer perspective, if they want to consume bits that are only available in net8.0-wasm, they should be able to change their TFM to target net8.0-wasm and everything should continue to compile since they would be targeting a superset of net8.0.